crates/api_common/src/request.rs

```diff
@@ -65,95 +65,84 @@ pub async fn fetch_link_metadata(url: &Url, context: &LemmyContext) -> LemmyResu
-
-    // Decide if we are allowed to generate local thumbnail
-    let allow_sensitive = local_site_opt_to_sensitive(&local_site);
-    let allow_generate_thumbnail = allow_sensitive || !post.nsfw;
+) -> LemmyResult<()> {
+  let metadata = match &post.url {
+    Some(url) => fetch_link_metadata(url, &context).await.unwrap_or_default(),
+    _ => Default::default(),
+  };
 
-    // Use custom thumbnail if available and its not an image post
-    let thumbnail_url = if !is_image_post && custom_thumbnail.is_some() {
-      custom_thumbnail
-    }
-    // Use federated thumbnail if available
-    else if federated_thumbnail.is_some() {
-      federated_thumbnail
+  let is_image_post = metadata
+    .content_type
+    .as_ref()
+    .is_some_and(|content_type| content_type.starts_with("image"));
+
+  // Decide if we are allowed to generate local thumbnail
+  let allow_sensitive = local_site_opt_to_sensitive(&local_site);
+  let allow_generate_thumbnail = allow_sensitive || !post.nsfw;
+
+  let thumbnail_url = if is_image_post {
+    if allow_generate_thumbnail {
+      match post.url {
+        Some(url) => generate_pictrs_thumbnail(&url, &context)
+          .await
+          .ok()
+          .map(Into::into),
+        None => None,
+      }
+    } else {
+      None
     }
-    // Generate local thumbnail if allowed
-    else if allow_generate_thumbnail {
-      match post
-        .url
-        .filter(|_| is_image_post)
-        .or(metadata.opengraph_data.image)
-      {
-        Some(url) => generate_pictrs_thumbnail(&url, &context).await.ok(),
+  } else {
+    // Use custom thumbnail if available and its not an image post
+    if let Some(custom_thumbnail) = custom_thumbnail {
+      proxy_image_link(custom_thumbnail, &context).await.ok()
+    } else if allow_generate_thumbnail {
+      match metadata.opengraph_data.image {
+        Some(url) => generate_pictrs_thumbnail(&url, &context)
+          .await
+          .ok()
+          .map(Into::into),
         None => None,
       }
     }
     // Otherwise use opengraph preview image directly
     else {
-      metadata.opengraph_data.image.map(Into::into)
-    };
-
-    // Proxy the image fetch if necessary
-    let proxied_thumbnail_url = proxy_image_link_opt_apub(thumbnail_url, &context).await?;
-
-    let form = PostUpdateForm {
-      embed_title: Some(metadata.opengraph_data.title),
-      embed_description: Some(metadata.opengraph_data.description),
-      embed_video_url: Some(metadata.opengraph_data.embed_video_url),
-      thumbnail_url: Some(proxied_thumbnail_url),
-      url_content_type: Some(metadata.content_type),
-      ..Default::default()
-    };
-    let updated_post = Post::update(&mut context.pool(), post.id, &form).await?;
-    if let Some(send_activity) = send_activity(updated_post) {
-      ActivityChannel::submit_activity(send_activity, &context).await?;
+      metadata.opengraph_data.image
     }
-    Ok(())
-  });
+  };
```

Comment:
Nested conditions can be avoided like this:

```
let image_url = if is_image_post {
  post.url
} else {
  metadata.opengraph_data.image
};

let thumbnail_url = if let (false, Some(url)) = (is_image_post, custom_thumbnail) {
  proxy_image_link(url, &context).await.ok()
} else if let (true, Some(url)) = (allow_generate_thumbnail, image_url) {
  generate_pictrs_thumbnail(&url, &context)
    .await
    .ok()
    .map(Into::into)
} else if !(is_image_post || allow_generate_thumbnail) {
  metadata.opengraph_data.image
} else {
  None
};
```

Reply:
We used to have something similar to that, which I removed in order to clean this up. Its less flat, but every condition is handled cleanly in these single ifs, rather than checking for multiple and possibly overlapping conditions, especially `is_image_post`

Reply:
Nested was an understatement. I really hate it. I had to look at it for a long time to understand it, and it's hard to tell if the resulting behavior is intended.

I still think the completely unnested version is best, but this would also be much better than the current one:

```
   let thumbnail_url = if is_image_post {
     if let (true, Some(url)) = (allow_generate_thumbnail, post.url) {
       generate_pictrs_thumbnail(&url, &context)
         .await
         .ok()
         .map(Into::into)
     } else {
       None
     }
   } else {
     // Use custom thumbnail if available and its not an image post
     if let Some(custom_thumbnail) = custom_thumbnail {
       proxy_image_link(custom_thumbnail, &context).await.ok()
     } else if let (true, Some(url)) = (allow_generate_thumbnail, metadata.opengraph_data.image) {
       generate_pictrs_thumbnail(&url, &context)
         .await
         .ok()
         .map(Into::into)
     } else {
       // Otherwise use opengraph preview image directly
       metadata.opengraph_data.image
     }
   };
```

Also, I realized that the last condition in the unnested version is redundant (images are not HTML and therefore have no opengraph tags), so it can be changed to this:

```
let image_url = if is_image_post {
  post.url
} else {
  metadata.opengraph_data.image
};

let thumbnail_url = if let (false, Some(url)) = (is_image_post, custom_thumbnail) {
  proxy_image_link(url, &context).await.ok()
} else if let (true, Some(url)) = (allow_generate_thumbnail, image_url) {
  generate_pictrs_thumbnail(&url, &context)
    .await
    .ok()
    .map(Into::into)
} else {
  metadata.opengraph_data.image
};
```

Reply:
Ya, the logic is def a bit hard to follow... I'll try this out. Its still doing several conditional checks on single lines rather than nesting them, which isn't as easy to follow IMO, but I'll try it.